compiler/optimizer/abstractinterpreter/AbsStackMachineState.hpp

```diff
@@ -0,0 +1,225 @@
+         _container(region)
+      {}
+
+   /**
+    * @brief Clone an op stack
+    * 
+    * @param region The memory region where the cloned op stack would like to allocated on.
+    * @return the op stack cloned.
+    */
+   TR::AbsOpStack* clone(TR::Region& region);
+
+   /**
+    * @brief Merge with another operand stack. This is in-place merge.
+    *
+    * @param other the operand stack to be merged with
+    */
+   void merge(TR::AbsOpStack* other);
+
+   /**
+    * @brief Push an abstract value to the operand stack.
+    * @note the abstract value to be pushed must be non-NULL.
+    *
+    * @param value the value to be pushed
+    */
+   void push(TR::AbsValue* value) { TR_ASSERT_FATAL(value, "Push a NULL value"); _container.push_back(value); }
+
+   /**
+    * @brief Peek the operand stack
+    *
+    * @return the abstract value
+    */
+   TR::AbsValue* peek() { TR_ASSERT_FATAL(size() > 0, "Peek an empty stack!"); return _container.back(); }
+
+   /**
+    * @brief Get and pop the value
+    *
+    * @return the abstract value
+    */
+   TR::AbsValue* pop();
+
+   /**
+    * @brief Set all the abstract values in this operand stack to the least precise ones
+    */
+   void setToTop();
+
+   bool empty()  { return _container.empty(); }
+   size_t size()  { return _container.size(); }
+  
+   void print(TR::Compilation* comp);
+
+   private:
+   TR::deque<TR::AbsValue*, TR::Region&> _container; 
+   };
+
+/**
+.* Abstract representation of the local variable array.
+ */
+class AbsLocalVarArray
+   {
+   public:
+   explicit AbsLocalVarArray(TR::Region &region) :
+         _container(region)
+      {}
+
+   /**
+    * @brief Clone an local var array
+    * 
+    * @param region The memory region where the cloned local var array would like to be allocated on
+    * @return the cloned local var array
+    */
+   TR::AbsLocalVarArray *clone(TR::Region& region);
+
+   /**
+    * @brief Merge with another abstract local var array. This is in-place merge.
+    *
+    * @param other The array to be merged with.
+    * @param vp value propagation
+    */
+   void merge(TR::AbsLocalVarArray* other);
+   
+   /**
+    * @brief Get the abstract value at index i.
+    *
+    * @param i the local var array index
+    * @return the abstract value
+    */
+   TR::AbsValue *at(uint32_t i);
+
+   /**
+    * @brief Set the abstract value at index i.
+    *
+    * @param i the local var array index
+    * @param value the abstract value to be set
+    */
+   void set(uint32_t i, TR::AbsValue* value);
+
+   /**
+    * @brief Set all the abstract values in this local var array to the least precise ones.
+    */
+   void setToTop();
+
+   size_t size() { return _container.size(); }
+   void print(TR::Compilation* comp);
+
+   private:
+   TR::deque<TR::AbsValue*, TR::Region&> _container;
+   };
+
+/**
+.* Abstract representation of the program state of the stack machine. 
+ */
+class AbsStackMachineState : public AbsState
```

Comment:
I'm not sure that saying `AbsStackMachineState` can represent the state of stack machine is accurate. It might be useful for some, but many stack machine languages need more than just an operand stack and local variable array.

Reply:
I am also not very familiar with other stack machine languages. So maybe putting this class into OpenJ9 will be a good idea?

Reply:
Moved to OpenJ9